src/event/utils.ts

```diff
@@ -25,10 +36,34 @@ export function defineEventHandler<
 >;
 export function defineEventHandler<
   Request extends EventHandlerRequest = EventHandlerRequest,
-  Response = EventHandlerResponse,
->(handler: EventHandler<Request, Response>): EventHandler<Request, Response> {
-  handler.__is_handler__ = true;
-  return handler;
+  Response = any,
+  Validator extends
+    ValidateFunction<EventHandlerRequest> = ValidateFunction<EventHandlerRequest>,
+  ValidatedRequest extends
+    EventHandlerRequest = Validator extends ValidateFunction<infer T>
+    ? T
+    : EventHandlerRequest,
+>(
+  handler: EventHandlerInput<Request, Response, Validator, ValidatedRequest>,
+): EventHandler<Request, Response> {
+  if (typeof handler === "function") {
+    return Object.assign(handler, { __is_handler__: true });
+  }
+  const wrapper: EventHandler<Request, any> = async (event) => {
+    if (handler.validate) {
+      await validateData(event, handler.validate);
+    }
+    for (const hook of handler.before || []) {
+      await hook(event);
+    }
+    const result = await handler.handler?.(event);
+    for (const hook of handler.after || []) {
+      await hook(event);
```

Comment:
Would be nice to also pass `, { body }` to the hooks to intercept the response. (body, to be consistent with global hooks)